geomesa-features/geomesa-feature-avro-complex/src/main/scala/org/locationtech/geomesa/features/avro/complex/AvroCodecs.scala

```diff
@@ -0,0 +1,704 @@
+package org.locationtech.geomesa.features.avro.complex
+
+import java.nio.ByteBuffer
+import java.sql.Timestamp
+import java.{net, sql, util}
+import java.util.function.Function
+import java.util.{Calendar, Date, GregorianCalendar, UUID, ArrayList => JArrayList, Collection => JCollection, Collections => JCollections, HashMap => JHashMap}
+
+import com.vividsolutions.jts.geom.Geometry
+import com.vividsolutions.jts.io.{WKBReader, WKBWriter}
+import org.apache.avro.Schema
+import org.apache.avro.io.{Decoder, Encoder}
+import org.geotools.feature.`type`.AttributeDescriptorImpl
+import org.geotools.feature.{AttributeImpl, ComplexAttributeImpl, FeatureImpl}
+import org.geotools.filter.identity.FeatureIdImpl
+import org.geotools.gml3.GMLSchema
+import org.locationtech.geomesa.features.avro.complex.AvroSchemas.{FieldNameEncoder, _}
+import org.locationtech.geomesa.utils.geotools.RichAttributeDescriptors._
+import org.opengis.feature.`type`.{PropertyDescriptor, _}
+import org.opengis.feature.{Attribute, ComplexAttribute, Property}
+import org.opengis.filter.identity.FeatureId
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+object AvroCodecs {
+
+  val primitiveTypes =
+    List(
+      classOf[String],
+      classOf[java.lang.Integer],
+      classOf[Int],
+      classOf[java.lang.Long],
+      classOf[Long],
+      classOf[java.lang.Double],
+      classOf[Double],
+      classOf[java.lang.Float],
+      classOf[Float],
+      classOf[java.lang.Boolean],
+      classOf[Boolean]
+    )
+
+  abstract case class AvroCodec[T](schema:Schema) {
+    val clazz: Class[T]
+    def enc(t: T, enc: Encoder): Unit
+    def dec(dec: Decoder): T
+  }
+
+  trait CodecRegistry {
+    protected val registry = mutable.Map.empty[Class[_],mutable.LinkedHashMap[Schema.Type,AvroCodec[_]]]

```

Comment:
question on thread safety here on the mutable map...i may see it later but can this be accessed from multiple threads? if so we need a concurrent hash map or similar mixin